server/controllers/stock/lots.js

```diff
@@ -198,12 +201,80 @@ async function merge(req, res, next) {
       transaction.addQuery(deleteLot, [mergeLotUuid]);
     });
     await transaction.execute();
-    res.sendStatus(200);
+    if (!req.params.internal) {
+      res.sendStatus(200);
+    }
   } catch (error) {
     next(error);
   }
 }
 
+/**
+ * GET /lots/merge/auto
+ *
+ * @description
+ * Finds and merges all lots suitable for automatic merging
+ *  - To qualify, the lots must have the same inventory_uuid,
+ *    label, and expiration date.
+ */
+function autoMerge(req, res, next) {
+  // The first query gets the inventory UUID for each
+  // inventory article with duplicate lots (having the
+  // same label, inventory_uuid, and expiration_date).
+  // (Since these are grouped, only one of the several
+  // lots is given.)
+  const query1 = `
+  SELECT
+    BUID(l.uuid) AS uuid, l.label, l.expiration_date,
+    BUID(i.uuid) AS inventory_uuid, i.text as inventory_text,
+    COUNT(*) as num_duplicates
+  FROM lot l
+  JOIN inventory i ON i.uuid = l.inventory_uuid
+  GROUP BY label, inventory_uuid, expiration_date HAVING num_duplicates > 1
+`;
+  // The second query gets all lots matching the label,
+  // inventory_uuid, and expiration dates
+  const query2 = `
+    SELECT
+      BUID(l.uuid) AS uuid, l.label, l.expiration_date,
+      BUID(i.uuid) AS inventory_uuid, i.text as inventory_text
+    FROM lot l
+    JOIN inventory i ON i.uuid = l.inventory_uuid
+    WHERE l.label=? AND i.uuid=? AND l.expiration_date=DATE(?)
+  `;
+  db.exec(query1, [])
+    .then((rows) => {
+      const numInventories = rows.length;
+      let numLots = rows.reduce((sum, row) => {
+        return sum + row.num_duplicates;
+      }, 0) - numInventories;
+      rows.forEach(async (row) => {
+        db.exec(query2, [row.label, db.bid(row.inventory_uuid), row.expiration_date])
+          .then((lots) => {
+            // Arbitrarily pick first lot and merge the duplicates into it
+            const masterLotUuid = lots[0].uuid;
+            const lotUuids = lots.reduce((list, elt) => {
+              if (elt.uuid !== masterLotUuid) {
+                list.push(elt.uuid);
+              }
+              return list;
+            }, []);
+            numLots = numLots + lotUuids.length - 1;
+
+            // Now merge this lot and its duplicates
+            merge({
```

Comment:
By the way, this wasn't obvious to me at first, but mixing Promises and async/await is totally fine.  In fact, async/await compiles to a Promise chain internally.

So, this works:
```js
async eleven() {

  // this variable will be 3
  const three = await Promise.resolve(3);
  
  // this will be 8
  const eight = await(Promise.resolve(5).then(five => five + three));
  
  // this will finally return 11
  return three + eight;
}
```

Reply:
You can make cute things like `sleep()` functions this way:

```js
// sleeps n seconds
function sleep(n) {
  return new Promise((resolve /*, reject */) => {
    setTimeout(() => {
      resolve();
    }, n * 1000);
  });
}

// mix async/await with a promise function
(async () => {
  console.log('Started!', Date.now());
  await sleep(3);
  console.log('Ended!', Date.now());
})();
```

Reply:
Thanks!